### PR TITLE
More AC Audio Artifacts

### PR DESCRIPTION
**Pull request by davidkelly**
AudioMixer has same issue transitioning from silence to encoded audio and back as we had in Agent audio code.  However, rather than throw away the state in the encoder, I now just encode one frame of 0s and send that before transitioning to silent frames.  This is basically just 0-padding the end of the sound, and seems to have done away with the clicks completely.  Done for both AudioMixer and Agent.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6a61762659e8578b3072d9189311a4d2cfe5fb93 available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-6a61762.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-6a61762.dmg

**Comment by kencooke** [comment]
I don't think a flush() method is the right approach here. IMHO a fast-path optimization (in this case, bypassing "silent" frames) should not be changing the result. Flush() discards the tail which can cause a click at the transition to silence. 

The codec has internal delay, so its output becomes silent one frame after the input becomes silent, which can be handled by a state machine. This is how the HRTF handles the transition to silence.

**Comment by ZappoMan** [comment]
I agree with Ken on this... you should be calling the codec API with a "lost packet" which is the same as silent...

**Comment by hifi-gustavo** [comment]
Build console output for commit: d00a73dde1c937e6e43ea54a1607f131fd4ee9c7 available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-d00a73d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-d00a73d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5e4f5391442ecf233934a744c213cc2e7fe22154 available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-5e4f539.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-5e4f539.dmg

**Comment by davidkelly** [comment]
# Test Plan
You want to have AC avatars make noise.  We were having issues with the `crowd-agent.js` and `summon.js` due to other issues which are being handled right now, so it maybe easier to use the `screamingAvatarAC.js` I've attached.  Just serve it up locally (say, `python -m SimpleHTTPServer 8888` from the directory that contains `screamingAvatarAC.js`.  Then, go to server settings and have a few instances of this script run.  I chose 10 due to my finite patience, but really even just 1 is good enough to hear or not hear the clicks.  
[screamingAvatarAC.txt](https://github.com/highfidelity/hifi/files/557182/screamingAvatarAC.txt)
(note I had to rename it .txt)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 292dd674335867cd070e09332aeb3dcca793f98e available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-292dd67.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-292dd67.dmg

File: assignment-client/src/audio/AudioMixer.cpp [hunk 1]
```diff
@@ -831,10 +832,14 @@ void AudioMixer::broadcastMixes() {
                         QString codecInPacket = nodeData->getCodecName();
                         mixPacket->writeString(codecInPacket);
 
-                        QByteArray decodedBuffer(reinterpret_cast<char*>(_clampedSamples), AudioConstants::NETWORK_FRAME_BYTES_STEREO);
                         QByteArray encodedBuffer;
-                        nodeData->encode(decodedBuffer, encodedBuffer);
-
+                        if ( mixHasAudio) {
```
Comment (zzmp) [inline assignment-client/src/audio/AudioMixer.cpp]: There is an extra space here.

File: assignment-client/src/audio/AudioMixerClientData.h [hunk 2]
```diff
@@ -76,7 +76,11 @@ class AudioMixerClientData : public NodeData {
         } else {
             encodedBuffer = decodedBuffer;
         }
+        // once you have encoded, you need to flush eventually.
+        _shouldFlushEncoder = true;
     }
+    void flushEncoder(QByteArray& encodedZeros);
```
Comment (zzmp) [inline assignment-client/src/audio/AudioMixerClientData.h]: This is not just flushing the encoder, but also encoding a flushing frame.

The name is confusing to me (as a reader), and makes it non-obvious why it needs a `QByteArray&` passed in.

File: assignment-client/src/Agent.h [hunk 3]
```diff
@@ -81,7 +81,7 @@ private slots:
 private:
     void negotiateAudioFormat();
     void selectAudioFormat(const QString& selectedCodecName);
-    void flushEncoder();
+    void flushEncoder(QByteArray& encodedZeros);
```
Comment (zzmp) [inline assignment-client/src/Agent.h]: This is not just flushing the encoder, but also encoding a flushing frame.

The name is confusing to me (as a reader), and makes it non-obvious why it needs a QByteArray& passed in.

File: assignment-client/src/Agent.cpp [hunk 4]
```diff
@@ -564,13 +558,20 @@ void Agent::processAgentAvatarAudio() {
             glm::quat headOrientation = scriptedAvatar->getHeadOrientation();
             audioPacket->writePrimitive(headOrientation);
 
-            QByteArray decodedBuffer(reinterpret_cast<const char*>(nextSoundOutput), numAvailableSamples*sizeof(int16_t));
             QByteArray encodedBuffer;
-            // encode it
-            if(_encoder) {
-                _encoder->encode(decodedBuffer, encodedBuffer);
+            if (_flushEncoder) {
+                // after sound is done playing, encoder has a bit of state in it,
```
Comment (zzmp) [inline assignment-client/src/Agent.cpp]: I think this comment belongs over the function declaration.

File: assignment-client/src/Agent.cpp [hunk 5]
```diff
@@ -564,13 +558,20 @@ void Agent::processAgentAvatarAudio() {
...
-            if(_encoder) {
-                _encoder->encode(decodedBuffer, encodedBuffer);
+            if (_flushEncoder) {
+                // after sound is done playing, encoder has a bit of state in it, 
+                // and needs some 0s to forget or you get a little click next time
+                // you play something.  So, basically 0-pad the end of the sound buffer
+                flushEncoder(encodedBuffer);
             } else {
-                encodedBuffer = decodedBuffer;
+                // encode it
```
Comment (zzmp) [inline assignment-client/src/Agent.cpp]: Belongs two lines down.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 52a14bf5c34d22c3c8101901618f62f0a60e9002 available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-52a14bf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-52a14bf.dmg

File: assignment-client/src/audio/AudioMixerClientData.h (hunk 2 above)
Comment (davidkelly) [inline assignment-client/src/audio/AudioMixerClientData.h]: Yea -- really encodeZeros or something like that.  Good point -- confusing function names add up after awhile.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e3a22618d1abaea263fc5f87cd0d0e6e1273e3be available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-e3a2261.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-e3a2261.dmg

**Comment by davidkelly** [comment]
please test this

**Comment by davidkelly** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: e3a22618d1abaea263fc5f87cd0d0e6e1273e3be available at: https://gist.github.com/9b9f1f619f68b1661d3080cdd9826d67<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-e3a2261.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8927-e3a2261.dmg

**Pull request merged**